rust/chains/hyperlane-sealevel/src/interchain_gas.rs

```diff
@@ -61,16 +236,82 @@ impl Indexer<InterchainGasPayment> for SealevelInterchainGasPaymasterIndexer {
-        Ok(vec![])
+        info!(
+            ?range,
+            "Fetching SealevelInterchainGasPaymasterIndexer InterchainGasPayment logs"
+        );
+
+        let mut payments = Vec::with_capacity((range.end() - range.start()) as usize);
+        for nonce in range {
+            if let Ok(sealevel_payment) = self.get_payment_with_sequence(nonce.into()).await {
+                let igp_account_filter = self.igp.igp_account;
+                if igp_account_filter == sealevel_payment.igp_account_pubkey {
+                    payments.push((sealevel_payment.payment, sealevel_payment.log_meta));
+                }
+            }
+        }
+        Ok(payments)
     }
 
     #[instrument(level = "debug", err, ret, skip(self))]
     async fn get_finalized_block_number(&self) -> ChainResult<u32> {
-        // As a workaround to avoid gas payment indexing on Sealevel,
-        // we pretend the block number is 1.
-        Ok(1)
+        get_finalized_block_number(&self.rpc_client).await
     }
 }
+
+#[async_trait]
+impl SequenceIndexer<InterchainGasPayment> for SealevelInterchainGasPaymasterIndexer {
+    async fn sequence_and_tip(&self) -> ChainResult<(Option<u32>, u32)> {
+        let program_data_account = self
+            .rpc_client
+            .get_account_with_commitment(&self.igp.data_pda_pubkey, CommitmentConfig::finalized())
+            .await
+            .map_err(ChainCommunicationError::from_other)?
+            .value
+            .ok_or_else(|| {
+                ChainCommunicationError::from_other_str("Could not find account data")
+            })?;
+        let program_data = ProgramDataAccount::fetch(&mut program_data_account.data.as_ref())
+            .map_err(ChainCommunicationError::from_other)?
+            .into_inner();
+        let payment_count = program_data
+            .payment_count
+            .try_into()
+            .map_err(StrOrIntParseError::from)?;
+        let tip = get_finalized_block_number(&self.rpc_client).await?;
+        Ok((Some(payment_count), tip))
+    }
+}
+
+#[test]
+fn test_unique_gas_payment_pubkey_offset() {
+    use borsh::BorshSerialize;
+    use hyperlane_sealevel_igp::accounts::GasPaymentData;
+    let expected_unique_gas_payment_pubkey = Pubkey::new_unique();
+
+    let gas_payment = GasPaymentAccount::new(
+        GasPaymentData {
+            sequence_number: 123,
+            igp: Default::default(),
+            destination_domain: Default::default(),
+            message_id: Default::default(),
+            gas_amount: Default::default(),
+            payment: Default::default(),
+            unique_gas_payment_pubkey: expected_unique_gas_payment_pubkey,
+            slot: Default::default(),
+        }
+        .into(),
+    );
+
+    let serialized = gas_payment.into_inner().try_to_vec().unwrap();
+    let sliced_unique_gas_payment_pubkey = Pubkey::new(
+        &serialized
+            [(UNIQUE_GAS_PAYMENT_PUBKEY_OFFSET - 1)..(UNIQUE_GAS_PAYMENT_PUBKEY_OFFSET + 32 - 1)],
```

Comment:
minus 1 because of the initialized bool being the first byte when `UNIQUE_GAS_PAYMENT_PUBKEY_OFFSET` is used in getProgramAccounts?

if so, a comment for future readers would be nice

Reply:
I think so, but it's not immediately apparent in the solana docs. Taking a closer look now